### PR TITLE
Add "Toggle Forbidden Store" to context menu, colorizes, and add preference

### PR DESCRIPTION
**Pull request by libraryaddict**
Adds the ability to toggle forbidden stores as shown in picture.
Colorizes the store a nice red if forbidden.
Adds a preference "showForbiddenStores" which defaults to false.

This also formats #369 it to be easier to read.

java_2021-12-18_12-45-09

### CODE REVIEW

File: src/net/sourceforge/kolmafia/request/MallPurchaseRequest.java [hunk 1]
```diff
@@ -49,22 +49,42 @@ public static void reset() {
...
-      }
-      list.add(shopIdString);
-      Preferences.setString("forbiddenStores", String.join(",", list));
+      return;
+    }
+
+    // forbidden is a fixed-sized list
+    ArrayList<String> list = new ArrayList<>(forbidden);
+
+    list.add(shopIdString);
```
Comment (gausie) [inline src/net/sourceforge/kolmafia/request/MallPurchaseRequest.java:66]: This looks wrong

File: src/net/sourceforge/kolmafia/request/MallPurchaseRequest.java [hunk 2]
```diff
@@ -157,6 +177,10 @@ public long getAvailableMeat() {
 
   @Override
   public String color() {
+    if (getForbiddenStores().contains(String.valueOf(this.shopId))) {
+      return "#c2002d";
```
Comment (gausie) [inline src/net/sourceforge/kolmafia/request/MallPurchaseRequest.java:181]: Can you get this colour from look and feel somehow to support those using dark modes

File: src/net/sourceforge/kolmafia/swingui/widget/ShowDescriptionList.java [hunk 3]
```diff
@@ -9,18 +9,10 @@
 import java.util.Map.Entry;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
-import javax.swing.JList;
-import javax.swing.JPopupMenu;
-import javax.swing.JSeparator;
-import javax.swing.SwingUtilities;
+import javax.swing.*;
```
Comment (gausie) [inline src/net/sourceforge/kolmafia/swingui/widget/ShowDescriptionList.java:12]: Don't do this

File: src/net/sourceforge/kolmafia/swingui/widget/ShowDescriptionList.java [hunk 4]
```diff
@@ -9,18 +9,10 @@
...
 import net.java.dev.spellcast.utilities.LockableListModel;
 import net.java.dev.spellcast.utilities.LockableListModel.ListElementFilter;
-import net.sourceforge.kolmafia.AdventureResult;
-import net.sourceforge.kolmafia.CreateFrameRunnable;
-import net.sourceforge.kolmafia.KoLConstants;
-import net.sourceforge.kolmafia.KoLmafiaCLI;
-import net.sourceforge.kolmafia.RequestThread;
-import net.sourceforge.kolmafia.StaticEntity;
+import net.sourceforge.kolmafia.*;
```
Comment (gausie) [inline src/net/sourceforge/kolmafia/swingui/widget/ShowDescriptionList.java:15]: Or this

File: src/net/sourceforge/kolmafia/swingui/widget/ShowDescriptionList.java [hunk 5]
```diff
@@ -31,14 +23,7 @@
...
 import net.sourceforge.kolmafia.preferences.Preferences;
-import net.sourceforge.kolmafia.request.AutoMallRequest;
-import net.sourceforge.kolmafia.request.AutoSellRequest;
-import net.sourceforge.kolmafia.request.CreateItemRequest;
-import net.sourceforge.kolmafia.request.PulverizeRequest;
-import net.sourceforge.kolmafia.request.PurchaseRequest;
-import net.sourceforge.kolmafia.request.UneffectRequest;
-import net.sourceforge.kolmafia.request.UseItemRequest;
-import net.sourceforge.kolmafia.request.UseSkillRequest;
+import net.sourceforge.kolmafia.request.*;
```
Comment (gausie) [inline src/net/sourceforge/kolmafia/swingui/widget/ShowDescriptionList.java:26]: Or this

**Review by gausie — COMMENTED** [review]
Would love some tests for this also. Rendering tests are pretty easy if you look into them!

File: src/net/sourceforge/kolmafia/request/MallPurchaseRequest.java [hunk 6]
```diff
@@ -157,6 +179,17 @@ public long getAvailableMeat() {
 
   @Override
   public String color() {
+    if (getForbiddenStores().contains(String.valueOf(this.shopId))) {
+      // Try get the color from look and feel
+      Color color = UIManager.getColor("InternalFrame.closePressedBackground");
```
Comment (gausie) [inline src/net/sourceforge/kolmafia/request/MallPurchaseRequest.java:184]: Suspicious that suspicious that you've had to take an intended background colour but if it works it works. You could even take the colour that we use for the error state on the compactsidepane

File: src/net/sourceforge/kolmafia/swingui/widget/ShowDescriptionList.java [hunk 7]
```diff
@@ -320,6 +323,28 @@ protected void executeAction() {
...
+    @Override
+    protected void executeAction() {
+      if (!(this.item instanceof PurchaseRequest)) {
+        return;
+      }
+
+      List<String> forbidden = MallPurchaseRequest.getForbiddenStores();
+      String storeId = ((PurchaseRequest) this.item).getFormField("whichstore");
+
+      if (storeId == null || !storeId.matches("[0-9]+")) {
```
Comment (gausie) [inline src/net/sourceforge/kolmafia/swingui/widget/ShowDescriptionList.java:336]: I think the code here could be simplified a bit by converting to an int here and catching numberformatexception

**Comment by gausie** [comment]
There have been a few changes from veracity underneath you but they should help

File: src/net/sourceforge/kolmafia/request/MallPurchaseRequest.java (hunk 6 above)
Comment (libraryaddict) [reply src/net/sourceforge/kolmafia/request/MallPurchaseRequest.java:188]: A bright cherry red sucks :P

Comment (gausie) [reply src/net/sourceforge/kolmafia/request/MallPurchaseRequest.java:188]: Hmm. Let's merge this but I think this is a minor bad decision

**Pull request merged**